File: code/day-3/js/04_loops.ts
/**
 * TypeScript Crash Course - Lesson 4: Loops (for, while, do-while)
 * ==================================================================
 * Learn to repeat actions efficiently
 */

// FOR LOOP - Iterate a specific number of times
console.log("=== BASIC FOR LOOP ===");
for (let i: number = 0; i < 5; i++) {  // 0 to 4
    console.log(`Count: ${i}`);
}
console.log();

// FOR LOOP WITH RANGE
console.log("=== FOR LOOP WITH RANGE ===");
for (let i: number = 1; i <= 10; i++) {  // 1 to 10
    console.log(`Number: ${i}`);
}
console.log();

// FOR LOOP WITH STEP
console.log("=== FOR LOOP WITH STEP ===");
// Print even numbers from 0 to 10
for (let i: number = 0; i <= 10; i += 2) {
    console.log(`Even number: ${i}`);
}
console.log();

// WHILE LOOP
// Repeats while a condition is true
console.log("=== WHILE LOOP ===");
let count: number = 1;
while (count <= 5) {
    console.log(`Count: ${count}`);
    count++;  // Important: update the counter!
}
console.log();

// DO-WHILE LOOP
// Executes at least once, then checks condition
console.log("=== DO-WHILE LOOP ===");
let num: number = 1;
do {
    console.log(`Number: ${num}`);
    num++;
} while (num <= 3);
console.log();

// BREAK STATEMENT
// Exit the loop prematurely
console.log("=== BREAK STATEMENT ===");
for (let i: number = 1; i <= 10; i++) {
    if (i === 5) {
        console.log("Breaking at 5");
        break;
    }
    console.log(`Number: ${i}`);
}
console.log();

// CONTINUE STATEMENT
// Skip the rest of the current iteration
console.log("=== CONTINUE STATEMENT ===");
for (let i: number = 1; i <= 5; i++) {
    if (i === 3) {
        console.log("Skipping 3");
        continue;
    }
    console.log(`Number: ${i}`);
}
console.log();

// NESTED LOOPS
console.log("=== NESTED LOOPS ===");
// Print a multiplication table
for (let i: number = 1; i <= 3; i++) {
    for (let j: number = 1; j <= 3; j++) {
        console.log(`${i} x ${j} = ${i * j}`);
    }
    console.log();  // Empty line after each row
}

// FOR...OF LOOP (for arrays)
console.log("=== FOR...OF LOOP ===");
const fruits: string[] = ["apple", "banana", "orange"];
for (const fruit of fruits) {
    console.log(`Fruit: ${fruit}`);
}
console.log();

// FOR...IN LOOP (for object properties)
console.log("=== FOR...IN LOOP ===");
const person = {
    name: "Alice",
    age: 25,
    city: "NYC"
};

for (const key in person) {
    console.log(`${key}: ${person[key as keyof typeof person]}`);
}
console.log();

// ARRAY forEach METHOD
console.log("=== ARRAY forEach METHOD ===");
const numbers: number[] = [1, 2, 3, 4, 5];
numbers.forEach((num: number, index: number) => {
    console.log(`Index ${index}: ${num}`);
});
console.log();

// PRACTICAL EXAMPLES
console.log("=== PRACTICAL EXAMPLES ===");

// Example 1: Sum of numbers from 1 to 10
let total: number = 0;
for (let i: number = 1; i <= 10; i++) {
    total += i;
}
console.log(`Sum of 1 to 10: ${total}`);
console.log();

// Example 2: Countdown timer
console.log("Countdown:");
let countdown: number = 5;
while (countdown > 0) {
    console.log(countdown);
    countdown--;
}
console.log("Blast off!");
console.log();

// Example 3: Find first number divisible by 7 and 5
for (let num: number = 1; num <= 100; num++) {
    if (num % 7 === 0 && num % 5 === 0) {
        console.log(`First number divisible by both 7 and 5: ${num}`);
        break;
    }
}
console.log();

// Example 4: Print only odd numbers
console.log("Odd numbers from 1 to 10:");
for (let i: number = 1; i <= 10; i++) {
    if (i % 2 === 0) {  // If even, skip
        continue;
    }
    process.stdout.write(`${i} `);
}
console.log("\n");

// Example 5: Factorial calculation
let factNum: number = 5;
let factorial: number = 1;
for (let i: number = 1; i <= factNum; i++) {
    factorial *= i;
}
console.log(`Factorial of ${factNum}: ${factorial}`);
console.log();

// Example 6: Pattern printing
console.log("Pattern:");
for (let i: number = 1; i <= 5; i++) {
    console.log("*".repeat(i));
}
console.log();

// Example 7: Reverse counting
console.log("Reverse count:");
for (let i: number = 5; i >= 1; i--) {
    console.log(i);
}
console.log();

// Example 8: Array iteration with index
console.log("=== ARRAY ITERATION WITH INDEX ===");
const colors: string[] = ["red", "green", "blue"];

console.log("Method 1: Traditional for loop");
for (let i: number = 0; i < colors.length; i++) {
    console.log(`  ${i}: ${colors[i]}`);
}
console.log();

console.log("Method 2: for...of with entries()");
for (const [index, color] of colors.entries()) {
    console.log(`  ${index}: ${color}`);
}
console.log();

// Example 9: Infinite loop prevention
console.log("=== SAFE LOOPING WITH MAX ITERATIONS ===");
let value: number = 10;
let maxIterations: number = 1000;
let iterations: number = 0;

while (value > 0 && iterations < maxIterations) {
    value -= 0.1;
    iterations++;
}
console.log(`Stopped after ${iterations} iterations`);

File: code/day-3/python/04_loops.py
"""
Python Crash Course - Lesson 4: Loops (for and while)
======================================================
Learn to repeat actions efficiently
"""

# FOR LOOP - Iterate over a sequence
print("=== BASIC FOR LOOP ===")
# Loop through a range of numbers
for i in range(5):  # 0 to 4
    print(f"Count: {i}")
print()

# FOR LOOP WITH RANGE
print("=== FOR LOOP WITH RANGE ===")
# range(start, stop, step)
for i in range(1, 11):  # 1 to 10
    print(f"Number: {i}")
print()

# FOR LOOP WITH STEP
print("=== FOR LOOP WITH STEP ===")
# Print even numbers from 0 to 10
for i in range(0, 11, 2):
    print(f"Even number: {i}")
print()

# WHILE LOOP
# Repeats while a condition is true
print("=== WHILE LOOP ===")
count = 1 # initialize the counter
while count <= 5: # condition to repeat the loop
    print(f"Count: {count}")
    count += 1  # Important: update the counter! # increment the counter
print()

# BREAK STATEMENT
# Exit the loop prematurely
print("=== BREAK STATEMENT ===")
for i in range(1, 11):
    if i == 5:
        print("Breaking at 5")
        break
    print(f"Number: {i}")
print()

# CONTINUE STATEMENT
# Skip the rest of the current iteration
print("=== CONTINUE STATEMENT ===")
for i in range(1, 6):
    if i == 3:
        print("Skipping 3")
        continue
    print(f"Number: {i}")
print()

# NESTED LOOPS
print("=== NESTED LOOPS ===")
# Print a multiplication table
for i in range(1, 4):
    for j in range(1, 4):
        print(f"{i} x {j} = {i * j}")
    print()  # Empty line after each row

# PRACTICAL EXAMPLES
print("=== PRACTICAL EXAMPLES ===")

# Example 1: Sum of numbers from 1 to 10
total = 0
for i in range(1, 11):
    total += i
print(f"Sum of 1 to 10: {total}")
print()

# Example 2: Countdown timer
print("Countdown:")
countdown = 5
while countdown > 0:
    print(countdown)
    countdown -= 1
print("Blast off!")
print()

# Example 3: Find first number divisible by 7 and 5
for num in range(1, 101):
    if num % 7 == 0 and num % 5 == 0:
        print(f"First number divisible by both 7 and 5: {num}")
        break
print()

# Example 4: Print only odd numbers
print("Odd numbers from 1 to 10:")
for i in range(1, 11):
    if i % 2 == 0:  # If even, skip
        continue
    print(i, end=" ")
print("\n")

# Example 5: Factorial calculation
number = 5
factorial = 1
for i in range(1, number + 1):
    factorial *= i
print(f"Factorial of {number}: {factorial}")
print()

number = 5
factorial = 1
while number > 0:
    factorial *= number
    number -= 1
print(f"Factorial of {number}: {factorial}")
print()

# Example 6: Pattern printing
print("Pattern:")
for i in range(1, 6):
    print("*" * i)
print()

# Example 7: While loop with sentinel value
print("Enter 'quit' to exit")
# Simulating user input with a list for demonstration
messages = ["Hello", "World", "quit"]
index = 0
while index < len(messages):
    message = messages[index]
    if message == "quit":
        print("Exiting...")
        break
    print(f"Message: {message}")
    index += 1

